Add tool use

JacksonKaunismaa opened the pull request:
Implements part of #11

abhimanyupallavisudhir commented:
Apologies for the delay; making comments as I go through it.

1) Tests are failing -- specifiically this seems to have broken caching (run e.g. `poetry run pytest -s tests/test_llm_utils.py::test_get_llm_response_returns_string[gpt-4o-mini]`). The first cell in playground also doesn't run for me, though it seems to have generated some output for you.

2) How do structured responses work now? Is `response_model` still supported? I can't tell immediately because the cache issue prevents stuff from running for me.

3) Having to hard code "btw you can use tools" in every prompt is not ideal.

JacksonKaunismaa commented:
> Apologies for the delay; making comments as I go through it.
> 
>     1. Tests are failing -- specifiically this seems to have broken caching (run e.g. `poetry run pytest -s tests/test_llm_utils.py::test_get_llm_response_returns_string[gpt-4o-mini]`). The first cell in playground also doesn't run for me, though it seems to have generated some output for you.
Should hopefully be fixed, though I think there is some issue with costly stripping out a `messages` kwarg that it expects to be list[dict[str, str]], but ours look like list[BaseMessage]. Not sure how we want to fix that.

>     2. How do structured responses work now? Is `response_model` still supported? I can't tell immediately because the cache issue prevents stuff from running for me.
It should still be supported, though I moved its definition so that it can be changed without recreating the llm, to match how tools are passed in. 

Also, I think the get_llm_response* things should be refactored so that in general we don't re-call `get_llm` each time, because that incurs significant overhead, especially for local models. 

>     3. Having to hard code "btw you can use tools" in every prompt is not ideal.

Fixed this too.

abhimanyupallavisudhir commented:
I pushed a change that feeds costly the messages converted to dict[str, str] form, though right now I'm just letting costly ignore ToolMessages.

I think that's fine because ToolMessages just contain the outputs of tool calls right? But I'm not sure how LangChain stores the tool list context -- is it in the system message? I know that in the standard list[dict[str,str]], it's stored like this:

```json
{
    "model": "gpt-3.5-turbo",
    "messages": [
        {
            "role": "user",
            "content": "Extract Jason is 25 years old"
        }
    ],
    "tools": [
        {
            "type": "function",
            "function": {
                "name": "UserDetail",
                "description": "Correctly extracted `UserDetail` with all the required parameters with correct types",
                "parameters": {
                    "properties": {
                        "name": {
                            "title": "Name",
                            "type": "string"
                        },
                        "age": {
                            "title": "Age",
                            "type": "integer"
                        }
                    },
                    "required": [
                        "age",
                        "name"
                    ],
                    "type": "object"
                }
            }
        }
    ],
    "tool_choice": {
        "type": "function",
        "function": {
            "name": "UserDetail"
        }
    }
}
```

I'm happy to merge now if everything else is fine, getting cost estimation perfect is not a priority.
